fix: Apply permlevel restrictions in Document, DatabaseQuery

gavindsouza opened the pull request:
### Changes

- Apply permlevel filters on Document REST & Client APIs
- Don't re-add popped Document fields in `as_dict`
- At most single get_meta query per get_list
- Check field permissions for requested and joined tables, pop the fields from the requested list which aren't accessible

### Misc Changes
- Added cleanup for Customizations added in Auto Repeat tests
- Fix bug in Database that printed None when DB clients raised SyntaxError
- Refactor test_reportview_get tests for better readability, split meaningfully

gavindsouza committed:
fix: Apply field permlevel for doc GET via REST

gavindsouza committed:
fix(rest): Delete doc attr if insufficient field permissions

gavindsouza committed:
perf: DatabaseQuery.prepare_args

* Re-use stripped str variable where possible
* Remove use of any + [], startswith to get rid of unnecessary evaluations

gavindsouza committed:
fix: Apply permlevel restrictions to DatabaseQuery

Allow reading only accessible fields for given user session if
ignore_permissions (get_all) is unset.

gavindsouza committed:
perf(db_query): Avoid re-fetching doctype meta

Store doctype meta in DatabaseQuery instance under `doctype_meta`
bypassing multiple get_meta calls and Redis/DB IO & serialization overheads.

gavindsouza committed:
fix: Fetch fields according to meta maintain order

gavindsouza committed:
fix(db_query): Maintain order of dict[/select] keys

* Reduce internals' mutating calls
* maintain order of fields as previous function
* Use performant f-strings over concat + formatting

gavindsouza committed:
fix(db_query): Apply permlevel checks on child/joined table queries

gavindsouza committed:
fix(db_query): Load doctype meta on demand not on init

gavindsouza committed:
fix: Pass parenttype in meta calls, handle count(*) type queries

gavindsouza committed:
fix(db_query): Parse SQL function calls to check if field is accessible

gavindsouza committed:
fix(doc): Maintain virtual df data in as_valid_dict

gavindsouza committed:
test: Update tests to handle missing attributes based on permlevel

gavindsouza committed:
fix(db_query): Allow standalone functions, rename get_permitted_fields

gavindsouza committed:
test: Add tests for permlevel handing in get_list

gavindsouza committed:
Merge branch 'develop' into permlevel-apis

gavindsouza committed:
test: Remove auto_repeat custom field in cleanup

gavindsouza committed:
fix(postgres): group_by in reportview, tests

Add approporiate group_by, order_by clauses in reportview and tests for
featureset compatibility with postgres

gavindsouza committed:
fix(db_query): Parse distinct field usages

ankush commented:
EPRNext: https://github.com/frappe/erpnext/actions/runs/3891260411

surajshetty3416 commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,62 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+		if self.flags.ignore_permissions:
+			return
+
+		available_fields = get_permitted_fields(doctype=self.doctype)
```
```suggestion
		permitted_fields = get_permitted_fields(doctype=self.doctype)
```
let's keep it consistent.

---
similarly update [this](https://github.com/frappe/frappe/pull/19533/files/19b728f51457cf29c7971c0d74b8dd9436ad0e05#diff-fb26f3edc73c37f97ceb65e510cb8714d0965e380173fe809628668f94b15c32R587) as well.

surajshetty3416 commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,62 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+				self.distinct = True
+				column = field.split(" ", 2)[1].replace("`", "")
+			else:
+				column = field.split(" ", 1)[0].replace("`", "")
```
Please add comments with conversion example

surajshetty3416 reviewed (changes requested):
Tested following things (to check for any side-effects for user with limited field access)
- API requests (list and doc)
- Report view (and report export)

Things seem to be working as expected.

gavindsouza committed:
refactor: Use `permitted` over `available` in variable naming

gavindsouza committed:
fix(db_query): Handle permlevel check cases clearer

- Split to utility functions for clarity
- Add example over code blocks
- Re-arrange blocks based on priority

gavindsouza committed:
test: Make test_child_table_join more resilient

gavindsouza committed:
test: Add test for linked table permission check

gavindsouza committed:
fix: Raise PermissionError when user doesnt have access to linked table

gavindsouza committed:
fix: Remove logic short circuit / dont reject fields easily

gavindsouza committed:
Merge branch 'develop' into permlevel-apis

gavindsouza commented:
ERPNext: https://github.com/frappe/erpnext/actions/runs/3931542659

gavindsouza committed:
Merge branch 'develop' into permlevel-apis

sagarvora commented on frappe/desk/reportview.py:
```diff
@@ -150,13 +150,13 @@ def setup_group_by(data):
 			if data.aggregate_on_field:
-				data.fields.append(f"`tab{data.aggregate_on_doctype}`.`{data.aggregate_on_field}`")
+				data.fields.append(column)
+			data.group_by += f", {column}"
```
This line seems to be causing an issue.

### Before

![Screenshot-2023-01-20-134639](https://user-images.githubusercontent.com/16315650/213648488-0b846994-37b0-4213-aedd-5898ee99b87f.png)


### After

![Screenshot-2023-01-20-134706](https://user-images.githubusercontent.com/16315650/213648499-4208d231-4f10-4408-b341-5fb80cb4d63b.png)

gavindsouza replied:
Should be fixed in 9efe84a644ab253820069d0b1511939cd5bbea4a; Fix was added because existing stuff wasn't playing well with Postgres. Unsure why https://github.com/frappe/frappe/pull/14424 (cc: @surajshetty3416) introduced the extra column in select but the test passes without it, and so do the reports on my MariaDB/Postgres sites that I tested locally.

gavindsouza committed:
fix(reportview): Remove aggregate_on_field added in fields

Essentially reverts changes added via
https://github.com/frappe/frappe/pull/14424 but works just the same
without the additional column :thonk:

The added column's data in mariadb was essentially garbage data. Wasn't
meaningful from what I could tell - couldn't play well with postgres
either

sagarvora commented on frappe/model/document.py:
```diff
@@ -675,14 +675,14 @@ def apply_fieldlevel_read_permissions(self):
-			if df.permlevel and not df.permlevel in has_access_to:
-				self.set(df.fieldname, None)
+			if df.permlevel and df.permlevel not in has_access_to:
+				delattr(self, df.fieldname)
```
`delattr` can fail is `self` doesn't have the attribute. problematic in many scenarios, including:
- if permlevel is set on a field like section break, which doesn't get set as an "attribute" (as in User doctype)

  ```python
  In [4]: doc.apply_fieldlevel_read_permissions()
  ---------------------------------------------------------------------------
  AttributeError                            Traceback (most recent call last)
  File <ipython-input-4-5a2c68b75259>:1, in <cell line: 0>()
  ----> 1 doc.apply_fieldlevel_read_permissions()
  
  File ~/Work/dev/apps/frappe/frappe/model/document.py:679, in Document.apply_fieldlevel_read_permissions(self)
      677 for df in self.meta.fields:
      678 	if df.permlevel and df.permlevel not in has_access_to:
  --> 679 		delattr(self, df.fieldname)
      681 for table_field in self.meta.get_table_fields():
      682 	for df in frappe.get_meta(table_field.options).fields or []:
  
  AttributeError: 'User' object has no attribute 'sb1'
  
  ```


- if this function happens to be called twice

sagarvora commented on frappe/model/base_document.py:
```diff
@@ -298,22 +299,25 @@ def get_valid_dict(
 						continue
 
-					if d[fieldname] is None and (options := getattr(df, "options", None)):
+					if d[fieldname] in {None, _DOC_DELETED_ATTR} and (options := getattr(df, "options", None)):
```
doesn't this code implicitly mean that virtual fields cannot have permlevel? It's not different than it was before, but it should be?

tldr: I think we need to check df.permlevel explicitly here.

gavindsouza replied:
Let's maintain this behavior in this PR - we should track permlevel for virtual fields in a separate issue.

gavindsouza replied:
Added a check. Wanted to iterate over permitted fields but that's causing all sorts of data inconsistency issues.

sagarvora commented on frappe/model/base_document.py:
```diff
@@ -349,6 +353,8 @@ def get_valid_dict(
 
 			if ignore_nulls and d[fieldname] is None:
 				del d[fieldname]
+			if not is_virtual_field and field_value is _DOC_DELETED_ATTR:
+				del d[fieldname]
```
could be unified as one `if` block

gavindsouza replied:
Separated because 4 conditions in 1 `if` was a bit hard to understand for me 😓 

---

Also, now that I look at this it should've been `elif` and not just `if`

sagarvora commented on frappe/model/base_document.py:
```diff
@@ -298,22 +299,25 @@ def get_valid_dict(
 		d = _dict()
 		for fieldname in self.meta.get_valid_columns():
+			field_value = getattr(self, fieldname, _DOC_DELETED_ATTR)
+
 			# column is valid, we can use getattr
-			d[fieldname] = getattr(self, fieldname, None)
+			d[fieldname] = field_value
```
if `field_value` is `_DOC_DELETED_ATTR` and field isn't virtual, we could just `continue` here instead of setting `d[fieldname]` only to delete it later?

Also important to note that if `sanitize` is False we're continuing below which means `_DOC_DELETED_ATTR` will become part of the dict and possibly make it non-serializable.

gavindsouza replied:
Removed `_DOC_DELETED_ATTR` and consequently the logic to remove unpermitted field keys from as_dict

sagarvora commented on frappe/model/meta.py:
```diff
@@ -530,6 +530,18 @@ def get_high_permlevel_fields(self):
 
 		return self.high_permlevel_fields
 
+	def get_permlevel_read_fields(self, parenttype=None, *, user=None):
```
Isn't this basically `get_readable_fields`? `get_permlevel_read_fields` makes it sound like it's permlevel-wise dict of fields.

I would rename the variable below as well.

gavindsouza replied:
fields suggests DocFields will be returned (seeing what other methods in the class suggest). Changing this to `get_permitted_fieldnames` with consideration with the suggestions below.

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,75 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+				column = strip_alias(column).replace("`", "")
+
+			if column == "*" and not in_function("*", field):
+				self.fields[i : i + 1] = permitted_fields
```
Could cause the "mutating list while looping" error 🤔

gavindsouza replied:
Don't think this affects objects other than dict & set. But if you come across this error, share context or suggest safer alternatives that we could use.

sagarvora replied:
let's assume it doesn't affect lists. you're still looping through those already permitted fields unnecessarily (`enumerate` returns an iterator).

alternatives:
- store index, do this at end of loop? (preferred)
- listify return value of enumerate
- do this at beginning of loop? you just need to check if the stripped value is "*" anyway - it can't have backticks or can't be in a function.

gavindsouza replied:
Implemented first suggestion via https://github.com/frappe/frappe/pull/19533/commits/2ae6b7f016506acea8b84de168ca13823e021d72

sagarvora commented on frappe/model/meta.py:
```diff
@@ -530,6 +530,18 @@ def get_high_permlevel_fields(self):
+				if df.permlevel in permlevel_access:
+					self.permlevel_read_fields.append(df)
+
+		return self.permlevel_read_fields
```
Perhaps this could be a **set** of **fieldnames**?
 - faster lookup
 - we only seem to be using fieldnames anyway

gavindsouza replied:
I'd expect the proper order of permitted fields from such an API. `List[DocField]` was returned for the sake of consistency / future-proofing - can do `List[str]` to better suit our use case.

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -1148,3 +1219,54 @@ def requires_owner_constraint(role_permissions):
+		meta_fields.remove("docstatus")
+
+	if meta.istable:
+		meta_fields.extend(["parent", "parenttype", "parentfield"])
```
use frappe.model.child_table_fields

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -1148,3 +1219,54 @@ def requires_owner_constraint(role_permissions):
+	else:
+		meta_fields.remove("idx")
+
+	return meta_fields + accessible_fields + optional_meta_fields
```
set for quicker lookup?

gavindsouza replied:
Need order to be maintained. Lookup + initialization gains/losses seem inconsequential here though

sagarvora replied:
why is order important here?

gavindsouza replied:
Document readability in REST / console

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -1148,3 +1219,54 @@ def requires_owner_constraint(role_permissions):
+
+def is_plain_field(field: str) -> bool:
+	for char in field:
+		if char in ("(", "`", ".", "'", '"', "*"):
```
(frozen)set constant?

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,75 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+				continue
+
+			# labels / pseudo columns or frappe internals
+			elif column[0] in {"'", '"', "_"}:
```
some users create fieldnames starting with underscore to prevent conflicts with core

gavindsouza replied:
Handled via https://github.com/frappe/frappe/pull/19533/commits/54ff630c77bd50005a88bd9d3fdca6f9c831643e

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,75 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+				ch_doctype = table.replace("`", "").replace("tab", "", 1)
+
+				if wrap_grave_quotes(table) in self.tables:
+					permitted_child_table_fields = get_permitted_fields(
+						doctype=ch_doctype, parenttype=self.doctype
+					)
```
should be a map instead of once per field?

---

permlevel of child table field in parent is checked in `frappe.permission.has_child_permission` so I think not checking it here should be fine.

gavindsouza replied:
> permlevel of child table field in parent is checked in frappe.permission.has_child_permission so I think not checking it here should be fine.

Permlevels don't seem to exist for Child Tables (at this point), but this block doesn't just deal with those but any joins being made in the query - like `Link.field` & `RandomTable.field`.

> should be a map instead of once per field?

Tried this at first. The current implementation is cheaper.

sagarvora commented on frappe/model/db_query.py:
```diff
@@ -550,6 +556,75 @@ def build_filter_conditions(self, filters, conditions, ignore_permissions=None):
+			elif "(" in field:
+				if "*" in field:
+					continue
+				elif any(x for x in permitted_fields if x in field):
```
seems problematic. what happens if I have a field called "apple" which is permitted, but "apple_pie" which isn't?

gavindsouza replied:
Removed this clause via https://github.com/frappe/frappe/pull/19533/commits/e41f005daad1da250524f31e03a808bbf5ac485e

gavindsouza committed:
fix: Check if attr exists before checking permlevel

gavindsouza committed:
fix(db_query): Set & use existing constants

gavindsouza committed:
refactor: Meta.get_permitted_fieldnames

* Remove older API that returned list[df]
* Rename fields in usage scope & Meta internals

gavindsouza committed:
fix(db_query): Permit optional_fields without checking in permitted
fields

gavindsouza committed:
fix: Handle * fields after field iterations

gavindsouza committed:
fix(db_query): Remove naive field in sql func check

gavindsouza committed:
refactor: get_valid_dict

* Util get_permitted_fields checks for valid columns instead of planned logic
* Remove virtual field from dict if not in permitted fields
* Remove reliance on sentinel object _DOC_DELETED_ATTR

gavindsouza committed:
Merge branch 'develop' into permlevel-apis

gavindsouza committed:
fix: Consider virtual fields in fields with values conditionally

gavindsouza committed:
fix(db_query): Check if params in sql fn call are all permitted access

gavindsouza commented:
ERPNext: https://github.com/frappe/erpnext/actions/runs/3931542659

gavindsouza committed:
fix(db_query): With as_list, pass NULL to maintain order

gavindsouza committed:
Merge branch 'develop' into permlevel-apis

gavindsouza commented:
~Not sure about the failing UI tests - they pass locally 🤔~

And poof

gavindsouza committed:
Merge branch 'develop' into permlevel-apis
